### PR TITLE
Feat(near-client): Chunk distribution via message bus

### PR DESCRIPTION
**Pull request by birchmd**
This PR implements the changes to nearcore proposed in https://github.com/near/nearcore/issues/10083

To summarize briefly here, the goal of this project is to reduce the latency experienced by RPC nodes through directly distributing chunks over a message bus. Validators eagerly push the chunks they produce on to this bus in addition to the peer-to-peer messages they send. When any node (validator or RPC) realizes it needs a chunk (e.g. because it is present in a new block) then it can check the message bus to see if it present there before trying a request over the peer-to-peer network.

Participation in chunk distribution via the message bus is entirely optional and disabled by default. This PR has no impact on nodes that are not participating in the new chunk distribution. To be clear: this PR is not a replacement for the existing chunk distribution logic via the peer-to-peer network; it is a secondary channel which should provide faster chunk distribution (on the happy path).

If invalid chunks are published to the message bus then nodes which receive them will request via the peer-to-peer network instead.

The details of how the chunks are posted to the message bus and what service exists for querying from the message bus are left intentionally abstract. These will be handled by separate a application (or applications). The nearcore code only sends HTTP requests to endpoints specified in config that is used by node operators to opt-in to this feature.

### CODE REVIEW

**Comment by wacban** [comment]
HI @birchmd, thanks for the proposal! I wasn't aware of this issue and it's great to see involvement from the community. I'll bring it up to the team for discussion and come back to you next week. 

First thoughts:
* I'm a tiny bit concerned that this is a centralized solution. I understand that this is opt in only but still we wouldn't want to build a dependence on it. 
* We will need to consider what are the alternatives and if there is any good distributed alternative. Perhaps we can just tweak the existing setup to broadcast a bit more. 
* We will need to double check the security. I don't think there should be any issues given we don't normally trust peers anyway but just to make sure I'd like to have a proper look. 
* The code looks nice, clean and well separated from the rest of the codebase which is nice.

**Comment by birchmd** [comment]
> I'm a tiny bit concerned that this is a centralized solution.

I completely understand the concern here. I think of this as an optimization of the happy path only, not any kind of protocol change. The decentralized peer-to-peer network is still the foundation on which everything works. This just gives an opportunity for faster message distribution when things are going well (which should be 99% of the time). The worst case scenario if something does go wrong with this is the latency of chunk distribution returns to what it was before this solution.

> tweak the existing setup to broadcast a bit more

I think it will be hard to compete with this using the peer-to-peer network. This is a solution optimized for this kind of message passing (multi-producer, multi-consumer with interest-based routing). I also wouldn't want to make the security of the core networking protocol worse by trying to improve its performance. I think it makes sense to have a separate network for performance.

> We will need to double check the security

+1 please review carefully.

> The code looks nice

❤️

**Comment by wacban** [comment]
We discussed it in core team and here are a few thoughts, comments and questions:

Alternatives - can you compare each to your proposal?
* Would it be possible for the RPC nodes to subscribe directly to a friendly validator instead? 

Comments for this solution:
* Can you do some experiments and share some relevant numbers e.g. what is the current latency and what is it using this solution?
* This solution should be made generic so that anyone can fully set it up. It should not rely on any closed source and we should share full instructions on how to configure it etc. 
* How do we make it the least intrusive into nearcore? Instead of coding it directly into nearcore can we implement a plug-in mechanism and keep the bus based implementation separate from nearcore? 

> I think it will be hard to compete with this using the peer-to-peer network. 

I think it boils down how much of a difference is there between the perfomance of the alternative solution. If we can implement a solution that is 10% slower but decentralized and safe to enable by default I would consider it the better alternative.  If it's a 2x difference that's a different story. 

Also JFYI we're making some changes to chunk processing and validation logic. In stateless validation chunks will have accompanying state witnesses and endorsements. Not sure if there are any changes necessary here but sharing just in case.

**Comment by bowenwang1996** [comment]
> This solution should be made generic so that anyone can fully set it up. It should not rely on any closed source and we should share full instructions on how to configure it etc.

Given that the message bus part is abstracted out, I don't think this PR necessarily introduces any point of centralization on its own. 

> How do we make it the least intrusive into nearcore? Instead of coding it directly into nearcore can we implement a plug-in mechanism and keep the bus based implementation separate from nearcore?

This is a good question. @birchmd what do you think?

**Comment by birchmd** [comment]
> Would it be possible for the RPC nodes to subscribe directly to a friendly validator instead?
> ...
> If we can implement a solution that is 10% slower but decentralized and safe to enable by default I would consider it the better alternative.

I don't think any kind of solution using the peer-to-peer network will be both performant and scalable. Essentially the trade-off is between latency (via pull messaging) vs bandwidth (actively pushing messages).

Currently chunks are distributed via pull (a node realizes they need some chunk and asks for it) and even with a direct connection to a peer that is guaranteed to have the chunk you will always have the round-trip latency of sending the pull message before getting the chunk in response. If we switch to a broadcast model (i.e. push) then we eliminate the round-trip latency, but traffic becomes an issue for the peer-to-peer network as the number of shards increases. This is why the base protocol is designed with pull messaging in the first place. You could have a push model that is more precise than broadcast by having some kind of subscription protocol, but I suspect validators will not want to open themselves up to having to spend even more bandwidth on messaging and therefore would reject subscription requests.

Even if we could switch chunk distribution to a push model, there is still the issue that subscribing to one validator is not enough. With one validator you would only automatically get the chunks that single validator produces (at least the changes to nearcore in this PR only forward chunks produced by the node as opposed to all chunks it receives). To consistently get all chunks from a shard you would need to subscribe to most of the validators assigned to that shard. That set of validators could change between epochs, requiring yet more complexity in the hypothetical subscription protocol we are discussing.

> Can you do some experiments and share some relevant numbers e.g. what is the current latency and what is it using this solution?

We have done some experiments and the solution should reduce latency by around 2 seconds. This number comes from measuring the time between an indexer receiving a block header for the first time and having received all chunks for that block. Additionally, on testnet we ran our own validator node with the patch to publish chunks it creates to a message bus and compared the time it took for an indexer to receive the chunk via the peer-to-peer network against the time for another indexer to receive that chunk via the message bus. This experiment suffered a little from lack of statistics because we could only measure time differences for chunks produced by our own validator.

The best experiment we could do would be to set up a sandbox network with programmable latencies and compare the peer-to-peer with the message bus in this setting. But this would be a larger undertaking and a significant delay for the project.

> This solution should be made generic so that anyone can fully set it up. It should not rely on any closed source and we should share full instructions on how to configure it etc.

All code for this project will be public and include documentation.

> How do we make it the least intrusive into nearcore? Instead of coding it directly into nearcore can we implement a plug-in mechanism and keep the bus based implementation separate from nearcore?
> This is a good question. @birchmd what do you think?

I think this PR is already minimally invasive to nearcore and is essentially a plug-in mechanism as-is. This PR only exposes sending and receiving chunk data at configurable HTTP endpoints. Anyone could design any other service that uses or produces chunk data and connect it to nearcore using this mechanism. For example, suppose a large validator wanted to use this to sync data between their production and fail-over nodes. This would be very easy; a simple HTTP server with two endpoints (post for accepting chunks from the production node and get for downloading those chunks to the fail-over node) would slot in perfectly with the changes in this PR.

This was the whole idea of only defining abstract HTTP interfaces in nearcore. It works for this project, while cleanly separating concerns and also opening up possibilities for future innovation.

File: chain/client/src/chunk_distribution_network.rs
Comment (pugachAG) [inline chain/client/src/chunk_distribution_network.rs:27]: nit: could you please specify what previous hash hash corresponds to `prev_block_hash`

File: chain/client/src/chunk_distribution_network.rs
Comment (pugachAG) [inline chain/client/src/chunk_distribution_network.rs:45]: nit: it is a bit weird to see `new` with return type `Option<Self>`, I suggest renaming to `from_config`

File: chain/client/src/chunk_distribution_network.rs
Comment (pugachAG) [inline chain/client/src/chunk_distribution_network.rs:95]: nit: just take owned `adapter` and make it caller responsibility to call `.clone()`

File: chain/client/src/chunk_distribution_network.rs
Comment (pugachAG) [inline chain/client/src/chunk_distribution_network.rs:41]: nit: would be nice to make those as private fields and expose underlying functionality as member functions, for example introduce `ChunkDistributionNetwork::enabled` to replace `chunk_distribution.config.enabled`

File: chain/client/src/client.rs
Comment (pugachAG) [inline chain/client/src/client.rs:1788]: nit: I guess there is no harm in having this `use` at the top of the file, this seems to be more conventional in nearcore codebase

File: chain/chunks/src/lib.rs
Comment (pugachAG) [inline chain/chunks/src/lib.rs:2029]: nit: do not interpolate error, just add it as a separate field: `warn!(target: "chunks", ?err, "Error processing partial encoded chunk");`

File: chain/client/src/client.rs
Comment (pugachAG) [inline chain/client/src/client.rs:1832]: nit:
```
       if let Some(chunk_distribution) = &self.chunk_distribution_network {
            if chunk_distribution.config.enabled {
                return crate::chunk_distribution_network::request_missing_chunks(
                    blocks_missing_chunks,
                    orphans_missing_chunks,
                    chunk_distribution.clone(),
                    &mut self.chain.blocks_delay_tracker,
                    &self.shards_manager_adapter,
                )
            }
        }
        self.p2p_request_missing_chunks(blocks_missing_chunks, orphans_missing_chunks);
```

**Review by pugachAG — APPROVED** [review]
LGTM with a bunch of optional improvement suggestions.

I confirm that this PR:
* doesn't introduce any protocol changes
* doesn't introduce any additional dependencies, interactions with centralised messaged bus are preformed via http
* doesn't change existing behaviour by default, the new feature needs to be explicitly enabled via config

File: chain/client/src/chunk_distribution_network.rs
Comment (birchmd) [reply chain/client/src/chunk_distribution_network.rs]: Addressed as part of https://github.com/near/nearcore/pull/10480/commits/2a4d18189d23cc4a4a8d72b248ac65b8d87006e0

File: chain/client/src/chunk_distribution_network.rs
Comment (birchmd) [reply chain/client/src/chunk_distribution_network.rs]: Addressed as part of https://github.com/near/nearcore/pull/10480/commits/2a4d18189d23cc4a4a8d72b248ac65b8d87006e0

File: chain/client/src/chunk_distribution_network.rs
Comment (birchmd) [reply chain/client/src/chunk_distribution_network.rs]: Addressed as part of https://github.com/near/nearcore/pull/10480/commits/2a4d18189d23cc4a4a8d72b248ac65b8d87006e0

File: chain/client/src/client.rs
Comment (birchmd) [reply chain/client/src/client.rs]: Addressed as part of https://github.com/near/nearcore/pull/10480/commits/2a4d18189d23cc4a4a8d72b248ac65b8d87006e0

File: chain/chunks/src/lib.rs
Comment (birchmd) [reply chain/chunks/src/lib.rs]: Addressed as part of https://github.com/near/nearcore/pull/10480/commits/2a4d18189d23cc4a4a8d72b248ac65b8d87006e0

File: chain/client/src/client.rs
Comment (birchmd) [reply chain/client/src/client.rs]: Addressed as part of https://github.com/near/nearcore/pull/10480/commits/2a4d18189d23cc4a4a8d72b248ac65b8d87006e0

**Comment by birchmd** [comment]
Thanks for the review @pugachAG ! I like all the suggestions you made and applied them in https://github.com/near/nearcore/pull/10480/commits/2a4d18189d23cc4a4a8d72b248ac65b8d87006e0

**Comment by bowenwang1996** [comment]
@birchmd what prevents this PR from getting merged?

**Comment by birchmd** [comment]
@bowenwang1996 nothing is blocking as far as I know. I'd love to see this merged!

**Comment by vikpande** [comment]
@wacban , can you please review and approve/comment. this is waiting for your approval to be merged. thanks

**Comment by pugachAG** [comment]
@vikpande Wac is OOO this week, but as far as I understand he is OK with this PR, so feel free to proceed with merging this.

**Comment by vikpande** [comment]
@birchmd , pls take it from here. good luck !

**Pull request merged**
